### PR TITLE
feat(database): implement order-independent equality for Reverts

### LINKED ISSUE #1813: Manually impl `PartialEq` for `Reverts`
In reth's invalid block hooks, we first compare the invalid block's `BundleState` to a re-executed version, then use `serde_json` to serialize `BundleState` and save it as JSON. Because the reverts are unsorted, this leads to equivalent `BundleState`s being seen as not equal. Instead, there should be a comparison API for `Reverts` that returns `true` if the reverts have the same contents, regardless of ordering.

### PR DESCRIPTION
**Pull request by hoank101**
Fixes #1813

### CODE REVIEW

**Comment by hoank101** [comment]
hi @Rjected please help me review it, thank!

File: crates/database/src/states/reverts.rs [hunk 1]
```diff
@@ -81,6 +82,39 @@ impl Reverts {
...
+                        self_map.remove(addr);
+                    }
+                    _ => return false,
+                }
+            }
+
+            if !self_map.is_empty() {
+                return false;
+            }
+        }
```
Comment (rakita) [inline crates/database/src/states/reverts.rs:113]: This does not look like a correct solution, there can be multiple reverts of same account.

Something like this is more preferred even if it is less performant:

```
let mut self_transition = self_transition.clone()
self_transition.sort()
other_transition.sort();
...
return self_transition == other_transition

```

Comment (hoank101) [reply crates/database/src/states/reverts.rs]: i just update, thank you

File: crates/database/src/states/reverts.rs [hunk 2]
```diff
@@ -81,6 +82,35 @@ impl Reverts {
...
+            return false;
+        }
+
+        for (self_transition, other_transition) in self.0.iter().zip(other.0.iter()) {
+            if self_transition.len() != other_transition.len() {
+                return false;
+            }
+            if self_transition.len() != other_transition.len() {
+                return false;
+            }
```
Comment (rakita) [inline crates/database/src/states/reverts.rs:97]:
```suggestion
```

duplicated

File: crates/database/src/states/reverts.rs [hunk 3]
```diff
@@ -81,6 +82,35 @@ impl Reverts {
...
+
+            let mut self_transition = self_transition.clone();
+            let mut other_transition = other_transition.clone();
+            // Sort both transitions
+            self_transition.sort_by(|(addr1, _), (addr2, _)| addr1.cmp(addr2));
+            other_transition.sort_by(|(addr1, _), (addr2, _)| addr1.cmp(addr2));
```
Comment (rakita) [inline crates/database/src/states/reverts.rs:103]: if addresses are equal, compare `AccountRevert`

File: crates/database/src/states/reverts.rs [hunk 4]
```diff
@@ -198,6 +236,115 @@ impl AccountRevert {
...
+impl Ord for RevertToSlot {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match (self, other) {
+            (RevertToSlot::Some(a), RevertToSlot::Some(b)) => a.cmp(b),
+            (RevertToSlot::Some(_), RevertToSlot::Destroyed) => Ordering::Less,
+            (RevertToSlot::Destroyed, RevertToSlot::Some(_)) => Ordering::Greater,
+            (RevertToSlot::Destroyed, RevertToSlot::Destroyed) => Ordering::Equal,
+        }
+    }
+}
```
Comment (Rjected) [inline crates/database/src/states/reverts.rs:308]: Can these `PartialOrd` / `Ord` just be derives on the types?

File: crates/database/src/states/reverts.rs [hunk 5]
```diff
@@ -198,6 +236,115 @@ impl AccountRevert {
...
+
+/// Implements total ordering for AccountRevert
+impl Ord for AccountRevert {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // First compare accounts
+        let account_ord = self.account.partial_cmp(&other.account);
+        if account_ord != Some(Ordering::Equal) {
+            return account_ord.unwrap();
```
Comment (Rjected) [inline crates/database/src/states/reverts.rs:316]: is this unwrap safe?

File: crates/database/src/states/reverts.rs [hunk 6]
```diff
@@ -198,6 +236,115 @@ impl AccountRevert {
...
+impl PartialOrd for AccountInfoRevert {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Implements total ordering for AccountInfoRevert
+impl Ord for AccountInfoRevert {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
```
Comment (Rjected) [inline crates/database/src/states/reverts.rs:249]: also wondering about this unwrap and whether or not we can just use `derive(PartialOrd, Ord)` on `AccountInfo Revert`

**Review by Rjected — DISMISSED** [review]

**Comment by hoank101** [comment]
hi @Rjected @rakita please help me review, thank

**Comment by hoank101** [comment]
Hi @rakita can you merge this PR?

**Review by rakita — APPROVED** [review]
lgtm! And thank you for waiting!

**Pull request merged**